addons/product/models/product_template.py

```diff
@@ -628,6 +628,10 @@ def _create_variant_ids(self):
 
         if variants_to_activate:
             variants_to_activate.write({'active': True})
+        if exclusion:
+            for variant in variants_to_activate:
+                if not self._is_combination_possible(variant.product_template_attribute_value_ids):
+                    variant.write({'active': False})
```

Comment:
As I said before, this kind of post-process is a worse case scenario where we need to iterate all variants a second time.
Ideally the variant should be put on the correct list immediately.

Also and to repeat it again, even if we kept this approach, it should be done before the operations are applied, not write true and then write false.

And also `_unlink_or_archive` should be called instead of setting active false manually.

But really it would be better if the 3 lists of operation were built correctly instead of having to adapt them.

Reply:
If I understand what you want, the solution would be something like:
```diff
--- a/addons/product/models/product_template.py
+++ b/addons/product/models/product_template.py
@@ -598,7 +598,8 @@ class ProductTemplate(models.Model):
@@ -598,19 +598,21 @@ class ProductTemplate(models.Model):
                 # For each possible variant, create if it doesn't exist yet.
                 for combination_tuple in all_combinations:
                     combination = self.env['product.template.attribute.value'].concat(*combination_tuple)
-                    if combination in existing_variants:
-                        current_variants_to_activate += existing_variants[combination]
-                    else:
-                        current_variants_to_create.append({
+                    is_combination_possible = self._is_combination_possible(combination)
+                    if is_combination_possible:
+                        if combination in existing_variants:
+                            current_variants_to_activate += existing_variants[combination]
+                        else:
+                            current_variants_to_create.append({
```

But this doesn't work because `_is_combination_possible` only works with active variants.

However, I'm still new and missing obviously something. And I don't know the way to virtually make the method `is_combination_possible` believe that the variant is active without creating/activating it.